docs/wiki/ALZ-Policies-Testing.md

```diff
@@ -0,0 +1,207 @@
+
+For ALZ, the focus is on testing Azure Policy definitions that have a DENY effect, as these can be very disruptive to organizations if a regression is introduced, and helps us improve the quality of the policies we are developing and deploying to production environments. The framework can be extended to test other policy effects, but this is not the focus of this framework.
+
+For authoring tests we standardized on using Az PowerShell native commands as much as possible as it is simpler to implement and read, however, there are circumstances where you will need to use REST APIs as not all features are exposed through Az PowerShell. To keep things simple, we have leverage the `Invoke-AzRestMethod` function that wraps the REST API calls and make it easier to use in Pester tests.
+
+### Prerequisites
+
+- [Pester](https://pester.dev/docs/introduction/installation)
+- [Az PowerShell Module](https://learn.microsoft.com/en-us/powershell/azure/install-azure-powershell?view=azps-11.0.0&viewFallbackFrom=azps-6.2.0)
+- [Invoke-AzRestMethod](https://learn.microsoft.com/en-us/powershell/module/az.accounts/invoke-azrestmethod?view=azps-11.0.0)
+
+### How it works
+
+The ALZ policy testing framework is designed to be used with GitHub Actions, but can be used with any CI/CD pipeline that supports PowerShell. The framework is designed to be used with the following workflow:
+
+1. A pull request is created to update a policy definition
+2. The pull request triggers a GitHub Action workflow
+3. The workflow runs the defined Pester tests against the policy definition
+4. The workflow reports the results of the tests back to the pull request checks
+5. The pull request is reviewed and handled based on the results of the tests
+
+### How to use it
+
+#### 1. Create a new GitHub Action workflow
+
+Create a new GitHub Action workflow in the `.github/workflows` folder of your repository. The workflow should be triggered on pull request events and should run on the `main` branch. The workflow should also allow being triggered manually to allow for testing of policies outside of pull requests.
+
+[Sample GitHub Action Workflow to run Policy tests](ALZ-Policies-Test-Workflow-Sample.md)
+
+#### 2. Create a new Pester test file
+
+Create a new Pester test file in the `tests/policy` folder of your repository. The test file should be named the same as the policy definition file it is testing, but with a `.tests.ps1` extension. For example, if the policy definition file is named `azurepolicy.json`, the test file should be named `azurepolicy.tests.ps1`.
+
+#### 3. Write the Pester tests
+
+Write the Pester tests in the test file. The tests should cover the following scenarios:
+
+- Conditions that should be true when the policy is evaluated, so it is compliant
+- Conditions that should be false when the policy is evaluated, so it is non-compliant
+
+It is important to test all the conditions evaluated in the policy. More on this below.
+
+### Where is the testing framework?
```

Comment:
Missing the workflow or an example one to call this.

Also the pre-reqs we have like the MG and policy defs deployed and subs moved.

Reply:
Good shout, added some more text around this. @jtracey93